merengueproj/docs/ref/settings.txt
.. _ref-settings:

==========================
Merengue specific settings
==========================

.. contents::
    :local:
    :depth: 1

Available settings
==================

Here's a full list of all available merengue settings. All these settings are
defined in `merengue/settings.py`_ and can be overriden in your project
settings.

.. _merengue/settings.py: http://tracpub.yaco.es/merengue/browser/trunk/merengueproj/merengue/settings.py


.. setting:: CATCH_STATUS_CODE

HTTP_STATUS_CODE_TEMPLATES
--------------------------

Default: {403: '403.html', 404: '404.html',}

When a HTTP status code (not 500) is found in response, you may want to use an
especific template to render. This is needed in example to catch all permission
error code (403 error).


.. setting:: SITE_FIXTURES

SITE_FIXTURES
-------------

Default: ``{}`` (Empty dictionary)

A dictionary mapping with fixtures to load for every application
It will executed after migrating (with south) models for every application.
Example:

.. code-block:: python

    SITE_FIXTURES = {
        'blog': ('blogs_data.xml', )
        'news': ('news_data.xml', 'news_users.json', )
    }

In this example, after migrate 'news' application Merengue will load
``news_data.xml`` and ``news_users.json``, located in one of `settings.FIXTURE_DIRS`_.

.. _settings.FIXTURE_DIRS: http://docs.djangoproject.com/en/1.3/ref/settings/#fixture-dirs


.. setting:: MERENGUE_URLS_PREFIX

MERENGUE_URLS_PREFIX
--------------------

Default: ``cms``

It will be the prefix for all URLs managed by merengue. Those URLs are defined
in `merengue/urls.py`_ file.

.. _`merengue/urls.py`: http://dev.merengueproject.org/browser/trunk/merengueproj/merengue/urls.py

In your project ``urls.py`` file, merengue URLs are included as follows:

.. code-block:: python

    # merengue URLs
    (r'^%s/' % settings.MERENGUE_URLS_PREFIX, include('merengue.urls')),

You can override this settings if you prefer other prefix than ``cms``.

.. setting:: URL_DEFAULT_LANG

URL_DEFAULT_LANG
----------------

Default: The ``LANGUAGE_CODE`` value

It will say to Merengue what URL will be used in internazionalized URLs.

For example, if you declare a in a plugin ``config.py`` file, a
``url_prefixes`` like that:

.. code-block:: python

    class PluginConfig(Plugin):
        # ... stuff
        url_prefixes = (
            ({'en': 'event', 'es': 'eventos'},
             'plugins.event.urls'),
        )

And your ``URL_DEFAULT_LANG`` is ``"es"``, the ``/eventos`` URL prefix will be
used.

.. setting:: CACHE_BACKEND

CACHE_BACKEND
-------------

Default: ``johnny.backends.locmem:///``

``CACHE_BACKEND`` is not a Merengue settings but a `Django one`_, but we mention
that settings because Merengue uses `Johnny Cache`_ as smart caching manager.

This implies you have to prepend ``johnny.backends.`` to your project caching
backends in order to get it works with ``johnny-cache`` (see
`Johnny Cache backends`_). Example:

.. code-block:: python

    CACHE_BACKEND = 'johnny.backends.memcached://127.0.0.1:11211/'

.. _Django one: http://docs.djangoproject.com/en/1.3/ref/settings/#cache-backend
.. _Johnny Cache: http://packages.python.org/johnny-cache/
.. _Johnny Cache backends: http://packages.python.org/johnny-cache/backends.html


.. setting:: TINYMCE_EXTRA_MEDIA

CACHE_SITE_FOR_ANONYMOUS
------------------------

Default: ``False``

``CACHE_SITE_FOR_ANONYMOUS`` activates the site template caching for anonymous requests. See :ref:`Optimizing a Merengue installation <topics-optimization>`.


TINYMCE_EXTRA_MEDIA
-------------------

Default: ``{ 'js': [], 'css': [] }``

Sometimes is useful for your site add extra CSS or Javascript files inside
Merengue HTML editor (TinyMCE). This would be useful to add your public styles
into HTML editor to be even more WYSIWYG.

You need include all media without ``MEDIA_URL`` prefix. This is an example:

.. code-block:: python

    TINYMCE_EXTRA_MEDIA = {
        'js': [],
        'css': ['themes/yourtheme/css/layout.css'],
    }


.. setting:: PLUGINS_DIR

PLUGINS_DIR
-----------

Default: ``plugins``

This setting defines when Merengue will look the plugins.


.. setting:: REQUIRED_PLUGINS

REQUIRED_PLUGINS
----------------

Default: ``('core', )``

A list with plugins required for the correct behaviour of your project. Those
plugins will be installed and activated when database is created and synced in
your project.


.. setting:: DETECT_NEW_PLUGINS

DETECT_NEW_PLUGINS
------------------

Default: ``False``

Detect automatically new plugins when manager access to plugins admin view.

If ``DETECT_NEW_PLUGINS`` is ``False``, registration of new plugins is done
executing ``register_new_plugins`` command.


.. setting:: DETECT_BROKEN_PLUGINS

DETECT_BROKEN_PLUGINS
---------------------

Default: ``True``

Detect automatically new plugins when manager access to plugins admin view.

If ``DETECT_BROKEN_PLUGINS`` is ``False``, broken plugins detection is done
executing ``mark_broken_plugins`` command.


.. setting:: USE_GIS

USE_GIS
-------

Default: ``False``

Activate GIS in Merengue. This must be defined previously to database creation.


.. setting:: MAP_FILTRABLE_MODELS

MAP_FILTRABLE_MODELS
--------------------

Default: ``('base.basecontent',)``

All models included will be displayed in a filter included in map view, when
GIS is activated.

COMPRESS
--------

Default: ``not DEBUG``

This setting activate/deactivate compression in Javascript and CSS files.
When deployment is done is recommended to activate this variable.

.. note::

    This variable will be activated by default if ``DEBUG`` is ``False``.

APPS_TO_TEST
------------

Default: ``('action', 'base', 'block', 'registry', 'perms', 'portal', 'pluggable', 'section', 'theming', 'viewlet', 'workflow', )``

This settings define the applications will be tested when the ``testmerengue`` command is executed.


merengueproj/CHANGES
Changes
=======

Version 0.8.0-rc1 (r5126)
-------------------------

* Permissions cache for speed up Merengue. See #2085.

* Optimized block rendering. See #2090.

* Refactored the permission system to generalize the roles acquisition from a section. See #2094.

* ``startplugin`` command to help developers to create plugins easily. See #2060.

* A new ``register`` plugin for user registration. See #2000.

* Enable inplace edit in the navigation block root elements. See #2055.

* A new ``Reviewer`` role for improving the publishing process flexibility. See #2093.

* Admin filters for installed or active plugins. See #2053.

* Automatic plugins detection is disabled by default, because there is a "detect new plugins" view. See #2051.

* Used the Django 1.3 message framework in the admin interface, to get latest features. See #2050.

* Direct link in the block configuration view to the admin view for this block. See #2061.

* Make more flexible the owners lookup for a content. See #2075.

* Disable by default to the user the permission editing capabilities in the "permission" tab. See #2138.

* Make the ``delete`` permission checking more flexible to be used in non ``BaseContent`` objects. See #2081.

* Unit tests in the block mechanism. See #2089.

* A new ``testmerengue`` command to test all the Merengue unit tests at once. See #2124.

* Make more homogeneous the forms CSS. See #73.

* Improved the selenium test runner to use frame buffer is needed. See #1999.

* Allow deleting of broken registered items like broken plugins. See #2130.

* Fixed a lot of bugs:

 * Reviewed permissions in related model admins. See #2042 and #2110.
 * Updating the open layers library to fix some FF3.0 errors. See #2084.
 * Custom permission in a content type are now visible in admin permissions view. See #2058.
 * Error adding non content blocks in the content scope. See #2077.
 * Make sure that contenttypes are sync after installing the plugin. This could cause weird issues. See #2052.
 * Fixed the detection of broken plugins in all circunstances. See #2054.
 * Workflows are loading without fixtures to avoid issues. See #2119.
 * Some translations. See #2045, #2097 and #2.
 * Others: #1772, #2063, #2064, #2066, #2067, #2042, #2074, #2076, #2077, #2080, #2082, #2086, #2088, #1744, #1989, #2092, #2096, #2098, #2100, #2101, #2102, #2103, #2104, #2105, #2106, #2109, #2112, #2114, #2045, #1923, #2117, #2118, #2120, #2121, #1982, #1972, #691, #2127, #2128, #2126, #2123, #2131, #2133.

Version 0.8.0-b2 (r4957)
------------------------

* Fixed a critical bug in 0.8.0-beta1 for newly created projects. See #2048.

* Fixed the workflow unit tests. See #1832.

* Improved Yaco theme layout. See #1990.

* Fixed some Selenium tests which was broken. See #1938.

* Hot link action for every content (disabled by default). See #2032.

* Fix a bug when adding blocks in a specific scope. See #2011.

* Other errors: #2037, #2038, #2039, #2036, #2040, #2041, #2043, #2046, #2039, #2047.

Version 0.8.0-b1 (r4880)
------------------------

* Django 1.3 compatibility. See #1568.

* Workflow system with management interface. See #990, #987, #988, #989.

* Added participants to contents and sections, in oposition of owners. See #1860.

* Documentation:

  * Improved English readability of documentation. See #377.
  * A new ``optimization`` section. See #1796.

* Activate a entire site cache with ``CACHE_SITE_FOR_ANONYMOUS`` setting. Also a ``invalidate cache`` panel. See #167.

* Block caching with management interface and cache invalidation in public view. See #1332, #1993, #1866 and #2007.

* More unit testing. See #1633, 

* Improvements to autoreports UI. See #1850, #1816.

* Better and nicer notifications. See #1841.

* The portal links categories can be configurable in the settings file. See #1858.

* Better l10n support in datetime format for spanish. See #1875.

* Developer can define the slug uniqueness in every inherited model. See #1873.

* Access by role for the portal links. See #1883.

* Add filtering to the block admin changelist. See #1806.

* Adding and removing blocks in the user interface. See #1913, #1923 and #2011.

* Mark the block container when adding blocks. See #1965.

* Changed the "content management" tab in admin interface. See #1750.

* Remove the ``Johnny-Cache`` fork because Django 1.3 support customs cache key prefixes. See #1758 and #1757.

* Included a new ``size`` parameter in the ``multimedia_slide_link`` templatetags. See #1756.

* Implemented a changelog view in admin interface. See #1778.

* Included italian translations. See #1803.

* All the managed contents types are orderable. See #1812.

* Multiples levels of portal links. See #1858.

* Support for fixed blocks (declared fixed by its developer). See #1056.

* Improve the linking to content from a menu item. See #1928.

* Add status to menus items. See #888.

* Allow to change the section of a content. See #1839.

* Implemented a "browse by type" admin view. See #1752.

* Allow to change active tabs in admin depending on user. See #1988.

* Inplace edit for image fields. See #1973.

* Update the Django Debug Toolbar. See #2001.

* New CSS class param for the blocks to gain look&feel flexibility. See #1977.

* Implemented access control by roles in menus and portal links. See #1846 and #1883.

* ``nondraggable`` and ``noncontained`` options for ``render_*_blocks`` templatetags. See #1895 and #1957.

* Work in permissions system to make more flexible the authorization control. See #2010, #2016, #2017, .

* Improvements in plugins:

 * New ``customportlet`` plugin to add customized chunks of HTML as you want. See #1761.

 * New ``piwik`` plugin, to collect statistics of accessing contents and sections using Piwik. See #1573.

 * New ``twitter`` plugin. See #1907.

 * New ``maps`` plugin, with a ``MapBlock`` block. See #1933 and #1941.

 * New ``googleanalytics`` plugin. See #1958.

 * Google search:


merengueproj/merengue/settings.py
# Copyright (c) 2010 by Yaco Sistemas
#
# This file is part of Merengue.
#
# Merengue is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Merengue is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with Merengue.  If not, see <http://www.gnu.org/licenses/>.

from os import path

ugettext = lambda s: s  # dummy ugettext function, as said on django docs

MERENGUEDIR = path.dirname(path.abspath(__file__))

# List  of  callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    #('django.template.loaders.cached.Loader', (
        'merengue.theming.loader.load_template_source',  # for enabling theme support in Merengue
        'django.template.loaders.filesystem.load_template_source',
        'django.template.loaders.app_directories.load_template_source',
        'django.template.loaders.eggs.load_template_source',
    #)),
)

TEMPLATE_CONTEXT_PROCESSORS = (
    'django.core.context_processors.auth',
    'django.core.context_processors.debug',
    'django.core.context_processors.i18n',
    'django.core.context_processors.media',
    'django.core.context_processors.request',
    'merengue.context_processors.all_context',
    'merengue.theming.context_processors.media',
    'merengue.section.context_processors.section',
)

TEMPLATE_DIRS = (
    path.join(MERENGUEDIR, 'templates'),
)

# merengue specific installed apps.
MERENGUE_APPS = (
    'merengue.registry',
    'merengue.pluggable',
    'merengue.multimedia',
    'merengue.perms',
    'merengue.workflow',
    'merengue.base',
    'merengue.section',
    'merengue.theming',
    'merengue.action',
    'merengue.block',
    'merengue.viewlet',
    'merengue.portal',
    'merengue.internallinks',
    'merengue.collection',
    'merengue.review',
    'merengue.uitools',
    'merengue.cache',
    'merengue.collab',  # Please, keep this application in last place cause it's used to know when activate plugins after migration
)

HTTP_STATUS_CODE_TEMPLATES = {
    403: '403.html',
    404: '404.html',
}

# Merengue installed apps
INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.admin',
    #'django.contrib.comments',
    'django_extensions',
    'template_utils',
    'debug_toolbar',
    'cmsutils',
    'transmeta',
    'transhette',
    'tagging',
    'sorl.thumbnail',
    'pagination',
    'inplaceeditform',
    'searchform',
    'inlinetrans',
    'mptt',
    'tinyimages',
    'rating',
    'captcha',
    'south',
    'threadedcomments',
    'autoreports',
    'formadmin',
    'johnny',
    'oot',
    'genericforeignkey',
    'oembed',
    'ajax_select',
    'notification',
    'compressor',
    'announcements',
) + MERENGUE_APPS

# Merengue specific middlewares.
MERENGUE_MIDDLEWARE_CLASSES = (
    'merengue.middleware.RemoveRandomAjaxParameter',
    'merengue.section.middleware.RequestSectionMiddleware',
    'merengue.section.middleware.DebugSectionMiddleware',
    'merengue.multimedia.middleware.MediaMiddleware',
    'merengue.middleware.SimplifiedLayoutMiddleware',
    'merengue.middleware.LocaleMiddleware',
    'merengue.middleware.HttpStatusCodeRendererMiddleware',
    'merengue.pluggable.middleware.ActivePluginsMiddleware',
    'django.middleware.csrf.CsrfResponseMiddleware',
)

# merengue usual middleware list. you can use this variable in your MIDDLEWARE_CLASSES project settings

PRE_MERENGUE_MIDDLEWARE_CLASSES = (
    'johnny.middleware.LocalStoreClearMiddleware',  # this has to be first
    'merengue.cache.middleware.UpdateAnonymousCacheMiddleware',
    #'cmsutils.middleware.ProfileMiddleware',  # remove comment if you want to profile your website
    'johnny.middleware.QueryCacheMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.middleware.transaction.TransactionMiddleware',
    'pagination.middleware.PaginationMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'cmsutils.middleware.AutomatizedTestingMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
)

POST_MERENGUE_MIDDLEWARE_CLASSES = (
    #'django.middleware.gzip.GZipMiddleware',
    'merengue.section.middleware.ResponseSectionMiddleware',
    'merengue.pluggable.middleware.PluginMiddlewaresProxy',
    'merengue.block.middleware.RenderBlockMiddleware',
    'merengue.cache.middleware.FetchFromAnonymousCacheMiddleware',
    'debug_toolbar.middleware.DebugToolbarMiddleware',
)

MIDDLEWARE_CLASSES = PRE_MERENGUE_MIDDLEWARE_CLASSES + MERENGUE_MIDDLEWARE_CLASSES + POST_MERENGUE_MIDDLEWARE_CLASSES

# merengue status list for contents workflow.
STATUS_LIST = (
    ('draft', ugettext('Draft')),
    ('pending', ugettext('Pending')),
    ('published', ugettext('Published')),
)
DEFAULT_STATUS = 'draft'

# max integer value (database safe)
MAX_INT_VALUE = 2147483647

# merengue default directory for plugins
PLUGINS_DIR = 'plugins'

# merengue required plugins in project (will be activated by default)
REQUIRED_PLUGINS = ('core', )

# if merengue will detect new plugins in file system
DETECT_NEW_PLUGINS = False

# if merengue will detect broken plugins
DETECT_BROKEN_PLUGINS = True

# The module to store session data
SESSION_ENGINE = 'merengue.backends.db'

# Roles which may be acquired by an user when visiting a content inside a section
ACQUIRE_SECTION_ROLES = False

# cache default settings
CACHE_MIDDLEWARE_SECONDS = 3600 * 24
CACHE_MIDDLEWARE_KEY_PREFIX = 'merengue'
CACHE_MIDDLEWARE_ANONYMOUS_ONLY = True
JOHNNY_MIDDLEWARE_KEY_PREFIX = 'merengue'

CACHE_PERMISSIONS = True
CACHE_MAX_KEY = 250

# Google API Key for localhost:8000
# http://code.google.com/apis/maps/signup.html
GOOGLE_MAPS_API_KEY = 'ABQIAAAAddxuy_lt2uAk9Y30XD3MJhQCULP4XOMyhPd8d_NrQQEO8sT8XBRRmJjQjU4qrycwOKb_v70y1h_1GQ'
GOOGLE_MAPS_API_VERSION = '3.x'
DEFAULT_LATITUDE = 36.851362
DEFAULT_LONGITUDE = -5.753321

# Pagination options
PAGE_VARIABLE = 'page'

# Default merengue options for tinyMCE
TINYMCE_MEDIA = None  # set to something like: MEDIA_URL + "cmsutils/js/widgets/tiny_mce/"
EXTRA_MCE = {
    'theme_advanced_buttons1': 'bold,italic,copy,paste,pasteword,underline,justifyleft,justifycenter,justifyright,justifyfull,bullist,numlist,outdent,indent',
}

TINYMCE_EXTRA_MEDIA = {
   'js': [],
   'css': [],
}

# Captcha default settings
CAPTCHA_SETTINGS = {
    'NUMBER_SWAP': {
                    'O': '0',
                    'Z': '2',
                    'S': '5',
                    'B': '8',
    },
    'CASE_SENSITIVE': False,
    'USE_EXTRA_IMAGE': False,
}

# Default parameters for suggestion box
SUGGESTION_BOX_EMAIL = 'info@foo.com'
CONTACT_SUGGESTIONBOX_PREFIX = 'SUGGESTION BOX'

# For transhette
ENABLE_TRANSLATION_SUGGESTIONS = False

# Map parameters
MAP_FILTRABLE_MODELS = (
    'base.basecontent',
)


def show_debug_toolbar(request):
    from django.conf import settings
    return settings.DEBUG_TOOLBAR


DEBUG_TOOLBAR_EXCLUDED_URLS = (
    r'^/tinyimages/',
)

DEBUG_TOOLBAR_CONFIG = {
    'INTERCEPT_REDIRECTS': False,
    'SHOW_TOOLBAR_CALLBACK': show_debug_toolbar,
}


# ajax_select
AJAX_LOOKUP_CHANNELS = {
    'perms_user': ('merengue.perms.lookups', 'UserLookup'),
    'perms_group': ('merengue.perms.lookups', 'GroupLookup'),
    'base_content': ('merengue.base.lookups', 'ContentLookup'),
    'user': ('merengue.base.lookups', 'UserLookup'),
    'section_contentlink': ('merengue.section.lookups', 'ContentLinkLookup'),
}

JQUERY_BASE_MEDIA = 'merengue/js/'

# Customization Comment app for merengue, feedback
# COMMENTS_APP = 'plugins.feedback'

MENU_PORTAL_SLUG = 'portal_menu'

# Dictionary with fixtures to load for every application
# After migrating (with south) 'foo' application Merengue will load
# every SITE_FIXTURES['foo'] files
# Syntax: {
#    'app_name': ('fixture_to_load1', 'fixture_to_load2', ...),
#    ...}
SITE_FIXTURES = {}

SERIALIZATION_MODULES = {
    "xml": "merengue.xml_serializer",
}

# Prefix for all merengue URLs
MERENGUE_URLS_PREFIX = 'cms'

# Login and logout settings
LOGIN_URL = '/%s/login/' % MERENGUE_URLS_PREFIX
LOGOUT_URL = '/%s/logout/' % MERENGUE_URLS_PREFIX
LOGIN_REDIRECT_URL = '/'
LOGOUT_REDIRECT_URL = None

# Test settings
TEST_RUNNER = 'merengue.test.run_tests'
APPS_TO_TEST = (
    'action',
    'block',
    'registry',
    'perms',
    'portal',
    'pluggable',
    'section',
    'theming',
    'viewlet',
    'workflow',
    'base',
)

# Merengue manage file
MANAGE_FILE = 'manage.py'

# sys.executable
# /usr/bin/python
SYS_EXECUTABLE = None

JOHNNY_TABLE_BLACKLIST = ('south_migrationhistory', 'django_content_type', 'thumbnail_kvstore', 'plugins_registeredplugin', )

# Path for translation catalogs search
LOCALE_PATHS = (
    path.join(MERENGUEDIR, 'locale'),
)

# Theme default preview image. The real path would be path.join(MEDIA_URL, DEFAULT_PLUGIN_PREVIEW)
DEFAULT_THEME_PREVIEW = 'merengue/img/preview_merengue.png'

STATIC_ROLES = [u'Anonymous User', u'Owner']

STATIC_WORKFLOW_DATA = {
    'states': [u'draft', u'published'],
    'transitions': [u'set-as-pending', u'publish', u'set-as-draft'],
}

# inplaceeditform
ADAPTOR_INPLACEEDIT = {'textarea': 'merengue.uitools.fields.AdaptorTinyMCEField',
                       'image': 'merengue.uitools.fields.AdaptorImageThumbnailField'}

# Plugin default preview image. The real path would be path.join(MEDIA_URL, DEFAULT_PLUGIN_PREVIEW)
DEFAULT_PLUGIN_PREVIEW = path.join('merengue', 'img', 'preview_merengue.png')

# Num element per page in collections
NUM_ELEM_PER_PAGE_DEFAULT = 10

# Search in genericforeingkey model admins
SEARCH_MODELADMIN = True

# Settings to autoreports
AUTOREPORTS_FUNCTIONS = True
AUTOREPORTS_INITIAL = True
AUTOREPORTS_I18N = True
AUTOREPORTS_SUBFIX = True

# Default workflow
DEFAULT_WORKFLOW = 1

# Portal link categories
PORTAL_LINK_CATEGORIES = (
    ('primary', ugettext('Primary link')),
    ('secondary', ugettext('Secondary link')),
)

PORTAL_LINK_POSITIONS = {
    'primary': 'header',
    'secondary': 'header',
}
